quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs

```diff
@@ -0,0 +1,630 @@
+
+/// A struct holding DateTime field options.
+#[derive(Clone, Serialize, Deserialize, Derivative)]
+#[derivative(Debug, Eq, PartialEq)]
+#[serde(deny_unknown_fields)]
+#[serde(from = "QuickwitDateOptionsDeser")]
+pub struct QuickwitDateOptions {
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// Accepted input formats.
+    #[serde(default = "default_input_formats")]
+    #[serde(serialize_with = "serialize_input_formats")]
+    #[serde(deserialize_with = "deserialize_input_formats")]
+    pub input_formats: BTreeSet<DateTimeFormat>,
+
+    /// Internal storage precision.
+    #[serde(default)]
+    pub precision: DatePrecision,
+
+    #[serde(default = "default_as_true")]
+    pub indexed: bool,
+
+    #[serde(default = "default_as_true")]
+    pub stored: bool,
+
+    #[serde(default)]
+    pub fast: bool,
+
+    /// A handle on DateTime parsing functions.
+    #[serde(skip)]
+    #[serde(default)]
+    #[derivative(PartialEq = "ignore")]
+    #[derivative(Debug = "ignore")]
+    parsers: DateTimeParsersHolder,
+}
+
+impl Default for QuickwitDateOptions {
+    fn default() -> Self {
+        Self {
+            description: None,
+            input_formats: default_input_formats(),
+            precision: DatePrecision::default(),
+            indexed: true,
+            stored: true,
+            fast: false,
+            parsers: DateTimeParsersHolder::default(),
+        }
+    }
+}
+
+impl QuickwitDateOptions {
+    pub fn parse_string(&self, value: String) -> Result<OffsetDateTime, String> {
+        self.parsers.parse_string(value)
+    }
+
+    pub fn parse_number(&self, value: i64) -> Result<OffsetDateTime, String> {
+        self.parsers.parse_number(value)
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields)]
+struct QuickwitDateOptionsDeser {
+    #[serde(default)]
+    pub description: Option<String>,
+    #[serde(default = "default_input_formats")]
+    #[serde(deserialize_with = "deserialize_input_formats")]
+    pub input_formats: BTreeSet<DateTimeFormat>,
+    #[serde(default)]
+    pub precision: DatePrecision,
+    #[serde(default = "default_as_true")]
+    pub indexed: bool,
+    #[serde(default = "default_as_true")]
+    pub stored: bool,
+    #[serde(default)]
+    pub fast: bool,
+}
+
+impl From<QuickwitDateOptionsDeser> for QuickwitDateOptions {
+    fn from(options: QuickwitDateOptionsDeser) -> Self {
+        let parsers = DateTimeParsersHolder::new(&options.input_formats);
+        QuickwitDateOptions {
+            description: options.description,
+            input_formats: options.input_formats,
+            precision: options.precision,
+            indexed: options.indexed,
+            stored: options.stored,
+            fast: options.fast,
+            parsers,
+        }
+    }
+}
+
+// Parser function types
+type StringDateTimeParser = Arc<dyn Fn(&str) -> Result<OffsetDateTime, String> + Sync + Send>;
```

Comment:
I don't think those function types are buying us anything. We could just iterate over the list of formats and get away with it. Less code, no dynamic dispatch. That would also remove the need for `QuickwitDateOptionsDeser`: